### PR TITLE
Sandrine and Henrike handing in project 16

### PR DESCRIPTION
**Pull request by HenrikeW**
Link to my deployed project: https://lostinmaze.netlify.app/

My reflections on how this week's project turned out: It was a lot of fun and very productive working together! We took a lot of help from previous projects but with that, everything went really smooth. The biggest struggle / most time consuming part was styling the page using grid and styled components and changing the styling (psition of buttons) depending on the information from the API.

Things I'd like to have clarified or explained in more detail: --

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,62 @@
...
+
+## Tech used
+
+- HTML5
+- CSS3
+- JavaScript ES6
+- React
+- Redux
+- styled components
+- Lottie animations
```
Comment (IngelaL) [inline README.md:58]: Great description of the game. You have described well what you have done and also what you haven't done.

File: code/public/index.html [hunk 2]
```diff
@@ -13,13 +16,14 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
-    <title>Technigo React App</title>
-  </head>
+  <title>Lost in maze</title>
+  <link rel="shortcut icon" href="%PUBLIC_URL%/icons/maze.ico">
```
Comment (IngelaL) [inline code/public/index.html:20]: 👍

File: code/src/components/ActionDescription.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+import { useSelector } from "react-redux";
+import styled from "styled-components/macro";
+
```
Comment (IngelaL) [inline code/src/components/ActionDescription.js:4]: Nice that you use /macro!

File: code/src/components/BackOrRestart.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+const ButtonContainer = styled.div`
+  grid-area: back;
+  justify-self: center;
+  align-self: center;
+  display: flex;
+  align-items: baseline;
+`;
+
+const GoBackText = styled(Text)`
```
Comment (IngelaL) [inline code/src/components/BackOrRestart.js:16]: Good job with reusable styling.

File: code/src/components/BackOrRestart.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    )
+  } else {
+    return (
+      <ButtonContainer>
+        <GoBackText>Does this feel wrong?</GoBackText>
+        <ActionButton onClick={onGoBack}>
+          Go back
+        </ActionButton>
+      </ButtonContainer>
+    )
```
Comment (IngelaL) [inline code/src/components/BackOrRestart.js:50]: We like the way you make use of the coordinates and that you offer the option of going back or restart the game.

File: code/src/components/Board.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    "back back back";
+`;
+
+export const Board = () => {
+  const stateUsername = useSelector((state) => state.game.username);
+  const isLoading = useSelector((state) => state.ui.isLoading);
+
+  return (
+    <>
+      {!isLoading && stateUsername && (
```
Comment (IngelaL) [inline code/src/components/Board.js:34]: You could comment more in the text, especially in this kind of code, to make it easier for a person who isn't familiar with the code.

File: code/src/components/Fireworks.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    animationData: animationData,
+    rendererSettings: {
+      preserveAspectRatio: "xMidYMid slice"
+    }
+  };
+
+  return (
+    <>
+      {coordinates === "1,3" && (
+        <Overlay>
```
Comment (IngelaL) [inline code/src/components/Fireworks.js:31]: Great use of the coordinates when to render specific components. And great that you made Lottie animation work.

File: code/src/lib/Styling.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+export const SmallText = styled(Text)`
+  font-size: 10px;
+  margin: 0 0 5px 0;
+
+  @media (min-width: 768px) {
+    font-size: 15px;
+    margin: 0 0 15px 0;
+  }
+`;
```
Comment (IngelaL) [inline code/src/lib/Styling.js:75]: Nice with reusable styling in a separate file.

**Review by IngelaL — APPROVED** [review]
**🔵 Blue Level (Minimum Requirements)**

- Your page should be responsive. - 👍 
- Use redux to store the current state of the game - Your application is very complex and you have put a lot of time and effort on this project. Great job!
- Focus on making the UX of your app good. Handle the response delay - We really like the possibility to go back and fourth and that the buttons are in the right directions. As first time experience, we feel that the instructions for each button are not very easy to spot. The text is a bit small  and doesn't coordinate with the rest of the design.
- Use thunks to wrap your API calls (but you don't have to if you don't want to). 👍 

In general we are very impressed with your work, we look forward to see you future project. 🥇

File: README.md [hunk 9]
```diff
@@ -1,13 +1,62 @@
...
+- a footer. 
+
+In the center of the page you see a fox as a little avatar. The buttons to move forward will appear above, below or next to the fox depending on the direction, like on a compass. 
+
+There's a specific start page where you only see an input field and the "Start the game"-button. On the last page, you don't have any direction buttons or descriptions as you can't move further. Because the API-requests take some time, there is a loader animation showing between the steps. There is a firework animation at the final step.
+
+## Limitations
+
+So far, the game has no local storage. The information about your progress will be stored in the API and as soon as you reload the page and reenter the game with the same username, you'll start at the same point where you left. But the information about the steps you have taken earlier won't be accessible, so you don't have the chance to go back. 
+If there had been more time, we would have implemented a storage. We would also have liked to use animations for the avatar to make the user experience more fun.
```
Comment (puj) [inline README.md:44]: As always, a wonderfully description readme. Love it.  
This section is more important than I can express.  Good job including it!

File: code/src/App.js [hunk 10]
```diff
@@ -1,9 +1,45 @@
...
+import styled from "styled-components/macro";
+
+import { Header } from "./components/Header";
+import { Start } from "./components/Start";
+import { Loading } from "./components/Loading";
+import { Board } from "./components/Board";
+import { Footer } from "./components/Footer";
+
+import { game } from "./reducers/game";
+import { ui } from "./reducers/ui";
```
Comment (puj) [inline code/src/App.js:13]: I like this structure of imports

File: code/src/App.js [hunk 11]
```diff
@@ -1,9 +1,45 @@
...
-      Find me in src/app.js!
-    </div>
+    <Provider store={store}>
+      <Grid>
+        <Header />
+        <Start />
+        <Loading />
+        <Board />
+        <Footer />
+      </Grid>
```
Comment (puj) [inline code/src/App.js:42]: Super clean

File: code/src/components/Board.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  grid-template-rows: 1fr 1fr 1fr 1fr 1fr 1fr 1fr;
+  grid-template-areas:
+    "status status status"
+    "description description description"
+    ". North ."
+    "West avatar East"
+    ". South ."
+    "instructions instructions instructions"
+    "back back back";
+`;
```
Comment (puj) [inline code/src/components/Board.js:26]: This could use a comment to clarify the intention

File: code/src/components/Fireworks.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+export const Fireworks = () => {
+  const coordinates = useSelector((state) => state.game.currentStep.coordinates);
+  const defaultOptions = {
+    loop: true,
+    autoplay: true,
+    animationData: animationData,
+    rendererSettings: {
+      preserveAspectRatio: "xMidYMid slice"
+    }
+  };
```
Comment (puj) [inline code/src/components/Fireworks.js:26]: Actually so cool 🎆

File: code/src/components/Start.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+import { game } from "../reducers/game";
+import { startGameFetch } from "../reducers/reusable";
+
+import { CenteredScreen, Button, StartText, Text } from "../lib/Styling";
+
+const InsertUsername = styled.input`
+  padding: 10px;
+  font-size: inherit;
+  font-family: inherit;
+`;
```
Comment (puj) [inline code/src/components/Start.js:14]: `InputUsername`
`UsernameInput`
`UsernameTextInput`

are some naming suggestions

File: code/src/reducers/game.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+      state.history = [...state.history, state.currentStep];
+      state.currentStep = action.payload;
+      state.stepCounter += 1;
+    },
+
+    oneStepBack: (state, action) => {
+      if (state.history.length > 0) {
+        state.currentStep = state.history[state.history.length - 1];
+        state.history = state.history.slice(0, state.history.length - 1);
+        state.stepCounter += 1;
```
Comment (puj) [inline code/src/reducers/game.js:33]: This is such a lovely detail.  Even a step backwards in a step in the right direction sometimes 👍

File: code/src/reducers/game.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+      } else {
+        alert("You need to take a step before you can go back")
```
Comment (puj) [inline code/src/reducers/game.js:35]: This never happens because you increment when entering the game.

**Review by puj — APPROVED** [review]
Very impressive work.  
I am a big fan of the arrows which make the game much more intuitive.  

Your variable/function naming is getting better and better.  Big A+ here. 
Create more constants for boolean logic.
Continue with the kickass readmes! 

💯

**Pull request closed without merging**